src/api/firebase.js

```diff
@@ -70,11 +74,62 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
 	});
 }
 
-export async function updateItem() {
-	/**
-	 * TODO: Fill this out so that it uses the correct Firestore function
-	 * to update an existing item. You'll need to figure out what arguments
-	 * this function must accept!
+/**
+ * Update an item in the user's list in Firestore.
+ * @param {string} listId The id of the list that has the items we're updating.
+ * @param {string} itemID The id of the list item we're updating.
+ * @param {boolean} checked The check state from ListItem.jsx.
+ */
+export async function updateItem(listId, itemId, checked) {
+	const listItemRef = doc(db, listId, itemId);
+	const listItemSnap = await getDoc(listItemRef);
+
+	const currentTotalPurchases = listItemSnap.data().totalPurchases;
+	let totalPurchases = currentTotalPurchases;
+
+	// the idea here is to store the old value of the `dateLastPurchased` property
+	// to a temporary variable called `originalDateLastPurchased`.
+	// We want this value so that we can revert back to it if the user un-checks an item.
+	let originalDateLastPurchased = listItemSnap
+		.data()
+		.dateLastPurchased.toDate();
+
+	console.log(originalDateLastPurchased);
+	// console output: a string date
+	// Fri Feb 03 2023 03:16:50 GMT-0800 (Pacific Standard Time)
+
+	// this is default setting `dateLastPurchased` as the old value (see line 93)
+	let dateLastPurchased = originalDateLastPurchased;
+
+	// if `checked` is true, `dateLastPurchased` is set to a new date.
+	// total purchases is also incremented.
+	if (checked === true) {
+		dateLastPurchased = new Date();
+		totalPurchases = currentTotalPurchases + 1;
+	} else {
+		totalPurchases = currentTotalPurchases - 1;
+		// we decrement total purchases
+		// we don't need to update `dateLastPurchased` in here, since default is already declared on line 102
+	}
+
+	// updateDoc updates the document's respective properties
+	await updateDoc(listItemRef, {
+		dateLastPurchased,
+		totalPurchases,
+	});
+
+	/** Note to mentors & collabies!
+	 *
+	 * This entire function currently only updates `totalPurchases`.
+	 * `dateLastPurchased` doesn't revert back to the original date when un-checked like we want to.
+	 *
+	 * The issue seems to arise from `listItemSnap` (line 85),
+	 * which only gets the most updated value of `dateLastPurchased`.
+	 * Therefore, the use of `originalDateLastPurchased` on line 93 doesn't work. It's always going to
+	 * have the updated date, not the old date.
+	 *
+	 * also, you don't see it here, but we also tried to experiment with date-fns, a dates JS library (this is a great library, btw!)
+	 * The issue still comes back to the list snapshot, though.
```

Comment:
Yes, lots of hard work has been dedicated to this.  This is apparent.  Spitballing: Could storing the date before being updated in another variable work?  Then calling it when unchecked?  There's got to be an easier way.  Look forward to hearing mentor input.  Again, big ups to you both for the effort this week.

Reply:
Thank you!! Haha yeah the storing the value was one thing we'd tried, but it ended up being tricky because the value of that variable would change when dateLastPurchased changed in Firestore 😅